crates/burn-compute/src/memory_management/dynamic.rs

```diff
@@ -0,0 +1,871 @@
+    Slice(SliceBinding),
+}
+
+/// The strategy defines the frequency at which deallocation of unused memory chunks should occur.
+#[derive(Debug)]
+pub enum MergingStrategy {
+    /// Once every n calls to reserve.
+    PeriodTick {
+        /// Number of calls to be executed before triggering the deallocation.
+        period: usize,
+        /// Current state. Should start at zero.
+        state: usize,
+    },
+    #[cfg(feature = "std")]
+    /// Once every period of time
+    PeriodTime {
+        /// Number of time before triggering the deallocation.
+        period: time::Duration,
+        /// Current state. Should start at now.
+        state: time::Instant,
+    },
+    /// Never deallocate.
+    Never,
+}
+
+/// The strategy defines when to reuse chunk with slices.
+#[derive(Debug)]
+pub enum SliceStrategy {
+    /// Never use slices.
+    Never,
+    /// Ratio needed before the chunk can be used as a slice. Between 0 and 1.
+    Ratio(f32),
+    /// When the reserved memory is at least {} bytes.
+    MinimumSize(usize),
+    /// When the reserved memory less than {} bytes.
+    MaximumSize(usize),
+}
+
+impl SliceStrategy {
+    /// If the chunk can be used with a slice.
+    pub fn can_use_chunk(&self, chunk_size: usize, reserved_size: usize) -> bool {
+        if chunk_size < reserved_size {
+            return false;
+        }
+
+        match self {
+            SliceStrategy::Never => false,
+            SliceStrategy::Ratio(ratio) => (reserved_size as f32 / chunk_size as f32) >= *ratio,
+            SliceStrategy::MinimumSize(bytes) => reserved_size >= *bytes,
+            SliceStrategy::MaximumSize(bytes) => reserved_size <= *bytes,
+        }
+    }
+}
+
+impl MergingStrategy {
+    /// Create a new strategy with the given period.
+    pub fn new_period_tick(period: usize) -> Self {
+        MergingStrategy::PeriodTick { period, state: 0 }
+    }
+
+    fn should_merge(&mut self) -> bool {
+        match self {
+            MergingStrategy::PeriodTick { period, state } => {
+                *state = (*state + 1) % *period;
+                *state == 0
+            }
+            #[cfg(feature = "std")]
+            MergingStrategy::PeriodTime { period, state } => {
+                if &state.elapsed() > period {
+                    *state = time::Instant::now();
+                    true
+                } else {
+                    false
+                }
+            }
+            MergingStrategy::Never => false,
+        }
+    }
+}
+
+#[derive(new, Debug)]
+struct Chunk {
+    storage: StorageHandle,
+    handle: ChunkHandle,
+    slices: Vec<SliceId>,
+}
+
+#[derive(new, Debug)]
+struct Slice {
+    storage: StorageHandle,
+    handle: SliceHandle,
+    // It is important to keep the chunk handle inside the slice, since it increases the ref count
+    // on the chunk id and make the `is_free` method return false until the slice is freed.
+    //
+    // TL;DR we can't only store the chunk id.
```

Comment:
Not true anymore since we don't return full chunks, we can remove the comment without changing the implementation.